test/jdk/jdk/internal/loader/URLClassPath/ClassPathUnusableURLs.java

```diff
@@ -0,0 +1,188 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.OutputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+
+import jdk.internal.loader.Resource;
+import jdk.internal.loader.URLClassPath;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+/*
+ * @test
+ * @bug 8344908
+ * @summary verify that when locating resources, the URLClassPath can function properly
+ *          without throwing unexpected exceptions when any URL in the classpath is unusable
+ * @modules java.base/jdk.internal.loader
+ * @run junit ClassPathUnusableURLs
+ */
+public class ClassPathUnusableURLs {
+
+    private static final Path SCRATCH_DIR = Path.of(".").normalize();
+    private static final String RESOURCE_NAME = "foo.txt";
+    private static final String SMILEY_EMOJI = "\uD83D\uDE00";
+
+    private static Path ASCII_DIR;
+    private static Path EMOJI_DIR;
+    private static Path JAR_FILE_IN_EMOJI_DIR;
+    private static int NUM_EXPECTED_LOCATED_RESOURCES;
+
+
+    @BeforeAll
+    static void beforeAll() throws Exception {
+        try {
+            EMOJI_DIR = Files.createTempDirectory(SCRATCH_DIR, SMILEY_EMOJI);
+        } catch (IllegalArgumentException iae) {
+            iae.printStackTrace(); // for debug purpose
+            // if we can't create a directory with an emoji in its path name,
+            // then skip the entire test
+            assumeTrue(false, "Skipping test since emoji directory couldn't be created: "
+                    + iae);
+        }
+        // successful creation of the dir, continue with the test
+        Files.writeString(EMOJI_DIR.resolve(RESOURCE_NAME), "hello");
+
+        ASCII_DIR = Files.createTempDirectory(SCRATCH_DIR, "test-urlclasspath");
+        Files.writeString(ASCII_DIR.resolve(RESOURCE_NAME), "hello");
+
+        // create a jar file containing the resource
+        JAR_FILE_IN_EMOJI_DIR = Files.createTempDirectory(SCRATCH_DIR, SMILEY_EMOJI)
+                .resolve("foo.jar");
+        final Manifest manifest = new Manifest();
+        manifest.getMainAttributes().putValue("Manifest-Version", "1.0");
+        try (final OutputStream fos = Files.newOutputStream(JAR_FILE_IN_EMOJI_DIR);
+             final JarOutputStream jaros = new JarOutputStream(fos, manifest)) {
+            final JarEntry jarEntry = new JarEntry(RESOURCE_NAME);
+            jaros.putNextEntry(jarEntry);
+            jaros.write("hello".getBytes(US_ASCII));
+            jaros.closeEntry();
```

Comment:
From a readability point of view, I think renaming "jaros" to "jos" would make it a bit easier to read. Also personally I would drop the uses of final from the try statement, only because there it adds noise when looking at the 3 resources in the statement.

Reply:
>  I think renaming "jaros" to "jos" would make it a bit easier to read.

Done.

> Also personally I would drop the uses of final from the try statement, only because there it adds noise when looking at the 3 resources in the statement.

Understood. I have updated the PR to remove them from the `try` statement.